sql-plugin/src/main/scala/com/nvidia/spark/rapids/CostBasedOptimizer.scala

```diff
@@ -0,0 +1,185 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeReference, Expression}
+import org.apache.spark.sql.execution.{ProjectExec, SortExec, SparkPlan}
+import org.apache.spark.sql.types.DataTypes
+
+object CostBasedOptimizer extends Logging {
+
+  // the intention is to make the cost model pluggable since we are probably going to need to
+  // experiment a fair bit with this part
+  private val costModel = new DefaultCostModel
+
+  /**
+   * Walk the plan and determine CPU and GPU costs for each operator and then make decisions
+   * about whether operators should run on CPU or GPU.
+   */
+  def optimize(plan: SparkPlanMeta[_]): (Double, Double) = {
+
+    // get the CPU and GPU cost of the child plan(s)
+    val (childCpuCosts, childGpuCosts) = plan.childPlans.map(optimize).unzip
+
+    // get the CPU and GPU cost of this operator
+    val (operatorCpuCost, operatorGpuCost) = costModel.applyCost(plan)
+
+    // calculate total (this operator + children)
+    val totalCpuCost = operatorCpuCost + childCpuCosts.sum
+    var totalGpuCost = operatorGpuCost + childGpuCosts.sum
+
+    // determine how many transitions between CPU and GPU are taking place
+    val numTransitions = plan.childPlans
+      .count(_.canThisBeReplaced != plan.canThisBeReplaced)
+
+    if (numTransitions > 0) {
+      if (plan.canThisBeReplaced) {
+        // transition from CPU to GPU
+        val transitionCost = numTransitions * costModel.transitionToGpuCost(plan)
+
+        if (operatorGpuCost + transitionCost > operatorCpuCost) {
+          plan.willNotWorkOnGpu("not worth moving to GPU just for this operator: " +
+              s"cpuCost=$operatorCpuCost; gpuCost=$operatorGpuCost; transitionCost=$transitionCost")
+        } else {
+          totalGpuCost += transitionCost
+        }
+
+      } else {
+        //  transition from GPU to CPU
+        totalGpuCost += numTransitions * costModel.transitionToCpuCost(plan)
+      }
+    }
+
+    logDebug(s"CBO: ${plan.wrapped.getClass} cpuCost=$totalCpuCost, gpuCost=$totalGpuCost, " +
+        s"numTransitions=$numTransitions")
+
+    if (totalGpuCost > totalCpuCost) {
+      forceOnCpuRecursively(plan, s"Forcing plan onto CPU: " +
+          s"cpuCost=$totalCpuCost; gpuCost=$totalGpuCost")
+      // reset the costs because everything below this point is on CPU now
+      (0, 0)
+    } else {
+      (totalCpuCost, totalGpuCost)
+    }
+  }
+
+  def forceOnCpuRecursively(plan: SparkPlanMeta[_], reason: String): Unit = {
+    if (plan.canThisBeReplaced) {
+      plan.willNotWorkOnGpu(reason)
+    }
+    plan.childPlans.foreach(plan => forceOnCpuRecursively(plan, reason))
```

Comment:
We need this to stop before it hits the end or if the last stage should not be on the GPU the entire query will be on the CPU.

Reply:
I'm not sure if I have completely addressed your concern here, but I changed it to only replace sections of the plan if the final operator could have been replaced.

Reply:
I see the limitations of this approach now that I am testing with TPC-DS and  I am rewriting this part. The current approach was too naive and we want to selectively force sections of the plan back onto CPU, not everything below a certain point.

Reply:
Right now we put everything back onto the CPU until we hit something that is already on the CPU.  I think that works fine for most of the stupid things we do now (like put something on the GPU just to rearrange columns).  But I don't think it will handle cases where something is more expensive on the GPU than the CPU and we may want to exclude just a part of the plan (instead of all of it).  But I think we can tackle that problem when we see it in practice.